plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -2030,6 +2039,78 @@ public void testIncorrectIcebergFileSizes()
         dropTable("test_iceberg_file_size");
     }
 
+    @Test
+    @Flaky(issue = "https://github.com/trinodb/trino/issues/5172", match = "Couldn't find operator summary, probably due to query statistic collection error")
+    public void testSplitPruningForFilterOnPartitionColumn()
+    {
+        String tableName = "nation_partitioned_pruning";
+
+        assertUpdate("DROP TABLE IF EXISTS " + tableName);
+
+        // disable writes redistribution to have predicatable number of files written per partition (one).
+        Session noRedistributeWrites = Session.builder(getSession())
+                .setSystemProperty("redistribute_writes", "false")
+                .build();
+
+        assertUpdate(noRedistributeWrites, "CREATE TABLE " + tableName + " WITH (partitioning = ARRAY['regionkey']) AS SELECT * FROM nation", 25);
+
+        // sanity check that table contains exactly 5 files
+        assertThat(query("SELECT count(*) FROM \"" + tableName + "$files\"")).matches("VALUES CAST(5 AS BIGINT)");
+
+        verifySplitCount("SELECT * FROM " + tableName, 5);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey = 3", 1);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 2", 2);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey < 0", 0);
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey >1 AND regionkey < 4", 2);
+
+        // TODO(https://github.com/trinodb/trino/issues/9309) should be 1
+        verifySplitCount("SELECT * FROM " + tableName + " WHERE regionkey % 5 = 3", 5);
+
+        assertUpdate("DROP TABLE " + tableName);
+    }
+
+    private void verifySplitCount(String query, int expectedSplitCount)
+    {
+        // using assertEventually here as the operator stats mechanism is known to be best-effort and some late-stage stats are sometimes not recorded
+        // (see https://github.com/trinodb/trino/issues/5172)
+        assertEventually(new Duration(10, TimeUnit.SECONDS), () -> {
```

Comment:
Shouldn't the `executeWithQueryId` be outside of the eventually loop?

Reply:
IDK to be honest. IIRC correctly we observed that sometimes statistics were never recorded for completed query; even after some time. So having it outside is safer if chance of success is big enough. 
The most safe approach would be to have longer assertEventually outside and shorter inside, but lets not do that unless we have to.